lib/http_clients/creditas/loan_api.ex

```diff
@@ -0,0 +1,103 @@
+defmodule HttpClients.Creditas.LoanApi do
+  @moduledoc false
+
+  alias HttpClients.Creditas.LoanApi
+
+  @spec get_by_key(Tesla.Client.t(), Key.t()) :: {:error, any} | {:ok, Loan.t() | nil}
+  def get_by_key(client, %LoanApi.Key{} = key) do
+    query = ["key.code": key.code, "key.type": key.type]
+
+    case Tesla.get(client, "/loans", query: query) do
+      {:ok, %Tesla.Env{status: 200, body: %{"items" => [attrs]}}} -> {:ok, build_loan(attrs)}
+      {:ok, %Tesla.Env{status: 200, body: %{"items" => []}}} -> {:ok, nil}
+      {:ok, %Tesla.Env{} = response} -> {:error, response}
+      {:error, reason} -> {:error, reason}
+    end
+  end
+
+  defp build_loan(attrs) do
+    key = to_struct(LoanApi.Key, attrs["key"])
+    contract = to_struct(LoanApi.Contract, attrs["contract"])
+    product = to_struct(LoanApi.Product, attrs["product"])
+    indexation = to_struct(LoanApi.Indexation, attrs["indexation"])
+
+    LoanApi.Loan
+    |> to_struct(attrs)
+    |> Map.put(:key, key)
+    |> Map.put(:contract, contract)
+    |> Map.put(:product, product)
+    |> Map.put(:indexation, indexation)
+    |> Map.put(:collaterals, build_collaterals(attrs["collaterals"]))
+    |> Map.put(:participants, build_participants(attrs["participants"]))
+    |> Map.put(:fees, build_fees(attrs["fees"]))
+    |> Map.put(:taxes, build_taxes(attrs["taxes"]))
+    |> Map.put(:interestRates, build_interest_rates(attrs["interestRates"]))
+    |> Map.put(:insurances, build_insurances(attrs["insurances"]))
+  end
+
+  defp build_collaterals(collaterals) do
+    Enum.map(collaterals, fn collateral ->
+      to_struct(LoanApi.Collateral, collateral)
+    end)
+  end
+
+  defp build_participants(participants) do
+    Enum.map(participants, fn participant ->
+      credit_score_attrs = participant["creditScore"]
+      credit_score = to_struct(LoanApi.CreditScore, credit_score_attrs)
+
+      LoanApi.Participant
+      |> to_struct(participant)
+      |> Map.put(:creditScore, credit_score)
+    end)
+  end
+
+  defp build_fees(fees) do
+    Enum.map(fees, fn fee -> to_struct(LoanApi.Fee, fee) end)
+  end
+
+  defp build_taxes(taxes) do
+    Enum.map(taxes, fn tax -> to_struct(LoanApi.Tax, tax) end)
+  end
+
+  defp build_interest_rates(interest_rates) do
+    Enum.map(interest_rates, fn interest_rate ->
+      to_struct(LoanApi.InterestRate, interest_rate)
+    end)
+  end
+
+  defp build_insurances(insurances) do
+    Enum.map(insurances, fn insurance -> to_struct(LoanApi.Insurance, insurance) end)
+  end
```

Comment:
```suggestion
  defp build_struct_list(kind, values) do: Enum.map(values, &to_struct(kind, &1))

  defp build_participants(participants) do
    Enum.map(participants, fn participant ->
      credit_score_attrs = participant["creditScore"]
      credit_score = to_struct(LoanApi.CreditScore, credit_score_attrs)

      LoanApi.Participant
      |> to_struct(participant)
      |> Map.put(:creditScore, credit_score)
    end)
  end
```
talvez assim? mta função fazendo a mesma coisa

Reply:
feito